src/lib.rs

```diff
@@ -100,45 +100,171 @@ pub mod visit;
 #[cfg(feature = "fold")]
 pub mod fold;
 
+////////////////////////////////////////////////////////////////////////////////
+
 #[cfg(feature = "parsing")]
-mod parsing {
-    use std::str::FromStr;
-
-    use super::*;
-    use synom::{Synom, ParseError};
-    use proc_macro2::TokenStream;
-
-    macro_rules! traits {
-        ($($ty:ident,)*) => ($(
-            impl From<TokenStream> for $ty {
-                fn from(stream: TokenStream) -> $ty {
-                    $ty::parse_all_unwrap(stream)
-                }
-            }
+pub use synom::ParseError;
 
-            impl FromStr for $ty {
-                type Err = ParseError;
+#[cfg(feature = "parsing")]
+use synom::{Synom, SynomBuffer};
 
-                fn from_str(s: &str) -> Result<Self, Self::Err> {
-                    $ty::parse_str_all(s)
-                }
+/// Parse tokens of source code into the chosen syn data type.
+///
+/// This is preferred over parsing a string because tokens are able to preserve
+/// information about where in the user's code they were originally written (the
+/// "span" of the token), possibly allowing the compiler to produce better error
+/// messages.
+///
+/// # Examples
+///
+/// ```rust,ignore
+/// extern crate proc_macro;
+/// use proc_macro::TokenStream;
+///
+/// extern crate syn;
+///
+/// #[macro_use]
+/// extern crate quote;
+///
+/// use syn::DeriveInput;
+///
+/// #[proc_macro_derive(MyMacro)]
+/// pub fn my_macro(input: TokenStream) -> TokenStream {
+///     // Parse the tokens into a syntax tree
+///     let ast: DeriveInput = syn::parse(input).unwrap();
+///
+///     // Build the output, possibly using quasi-quotation
+///     let expanded = quote! {
+///         /* ... */
+///     };
+///
+///     // Parse back to a token stream and return it
+///     expanded.parse().unwrap()
+/// }
+/// ```
+#[cfg(feature = "parsing")]
+pub fn parse<T, U>(tokens: U) -> Result<T, ParseError>
```

Comment:
> Ok so to make this testable I had to change this to using `Into<proc_macro2::TokenStream>` instead of `proc_macro::TokenStream`. The latter requires the context of the compiler to actually be created, effectively making `proc_macro` impossible to use in unit tests. (at least as constructed today)

If this is really only motivated by unit tests as Alex wrote, I would prefer to have a hidden function for the tests to use and stick with `proc_macro::TokenStream` for the public one. `syn::parse::<Expr, _>` is not nice.